pubsub/google/cloud/pubsub_v1/subscriber/_protocol/leaser.py

```diff
@@ -116,14 +138,14 @@ def maintain_leases(self):
             # we're iterating over it.
             leased_messages = copy.copy(self._leased_messages)
 
-            # Drop any leases that are well beyond max lease time. This
-            # ensures that in the event of a badly behaving actor, we can
-            # drop messages and allow Pub/Sub to resend them.
+            # Drop any leases that are beyond the max lease time. This ensures
+            # that in the event of a badly behaving actor, we can drop messages
+            # and allow the Pub/Sub server to resend them.
             cutoff = time.time() - self._manager.flow_control.max_lease_duration
             to_drop = [
-                requests.DropRequest(ack_id, item.size)
+                requests.DropRequest(ack_id, item.size, item.ordering_key)
                 for ack_id, item in six.iteritems(leased_messages)
-                if item.added_time < cutoff
+                if item.sent_time < cutoff
```

Comment:
Answering myself - seems like it's intended, since we expect the user to either `ack()` or `nack()` the message that both cause the message to be dropped from the leaser. If neither of these actions happen, it is correct to drop the message automatically upon lease expiration.

Reply:
Subquestion - what if message processing stalls for some reason and we never release any more messages, meaning that for some of the leased messages the lease expiry timer will not start? It could thus happen that we keep a (limited) number of messages in the leaser indefinitely, even those that might expire on the server side in the meantime (e.g. due to hitting the max lease duration limit).

If a subscriber is finally resumed at that point, wouldn't that cause expired messages to be released to the user?

Reply:
Yes, you answer is correct. They'd be dropped and re-delivered, to this subscriber or another one. This is especially good to have if the client is stuck in some way.

"If a subscriber is finally resumed at that point, wouldn't that cause expired messages to be released to the user?" -> That's fine since Pub/Sub has an "at least once" guarantee. Redelivery can happen for many other reasons in addition to the one you've pointed out.

Reply:
Thinking about it a bit more, this can already happen anyway - if message processing halts for a considerable amount of time and is then resumed, the messages that are dispatched to the user's callbacks might have expired in the meantime, and will be re-delivered by the server.